subsys/net/l2/ethernet/ethernet.c

```diff
@@ -174,10 +174,16 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 	struct ethernet_context *ctx = net_if_l2_data(iface);
 	struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
 	uint8_t hdr_len = sizeof(struct net_eth_hdr);
-	uint16_t type = ntohs(hdr->type);
+	uint16_t type;
 	struct net_linkaddr *lladdr;
 	sa_family_t family;
 
+	if (hdr == NULL) {
```

Comment:
`hdr` comes from `net_pkt_data(pkt)`. Are we sure that `pkt->buffer` is non-NULL? Maybe its safer to use `net_pkt_is_empty()`, that you introduced lately?

While at it I also wonder if we should check for `pkt->buffer->len >= hdr_len`.

Reply:
> Are we sure that `pkt->buffer` is non-NULL?

That pointer is/has been checked from day one in the network stack, I am pretty sure it is not NULL at this point. The problem was the pkt->buffer->data pointer that could have been NULL if we run out of bufs.
I think it is more intuitive to check the `hdr` here instead of net_pkt as we are later using the hdr variable.

> While at it I also wonder if we should check for `pkt->buffer->len >= hdr_len`.

Yes, it looks like we never check that there is enough bytes for Ethernet header. I will add this check.

Reply:
>The problem was the pkt->buffer->data pointer that could have been NULL if we run out of bufs.

If we run out of bufs (meaning net_buf structures), then `pkt->frags` (AKA `pkt->buffer`) is going to be NULL. I don't see how `pkt->buffer->data` can be NULL for net_buf_fixed type of net_bufs. What am I missing?

Reply:
> If we run out of bufs (meaning net_buf structures), then `pkt->frags` (AKA `pkt->buffer`) is going to be NULL. I don't see how `pkt->buffer->data` can be NULL for net_buf_fixed type of net_bufs. What am I missing?

I have been looking the tcp2 and net_buf code quite a lot lately, and it is a bit difficult to see all the interactions when the out-of-buffer situation happens. One thing I noticed that looks a bit suspicious, is that the net_buf refcounting is not atomic which could cause some issues or not.